Mathlib/Data/Finite/PermOf/Defs.lean

```diff
@@ -0,0 +1,276 @@
+    have hb := b.getElem_lt (hi := hi)
+    rw [Nat.lt_one_iff] at ha hb
+    exact ha.trans hb.symm
+
+instance : Finite (PermOf n) := Finite.of_injective
+  (fun a => (⟨a[·], a.getElem_lt⟩ : Fin n → Fin n)) <| fun a b => by
+    simp_rw [funext_iff, Fin.ext_iff, PermOf.ext_iff, Fin.forall_iff,
+      Fin.getElem_fin, imp_self]
+
+end Ext
+
+instance : One (PermOf n) where
+  one := PermOf.mk (Vector.range n) (Vector.range n)
+    (fun _ => by simp_rw [Vector.getElem_range, imp_self])
+    (by simp_rw [Vector.getElem_range, implies_true])
+
+section One
+
+@[simp]
+theorem getElem_one {i : ℕ} (hi : i < n) : (1 : PermOf n)[i] = i := Vector.getElem_range _
+
+instance : Inhabited (PermOf n) := ⟨1⟩
+
+@[simp]
+theorem default_eq : (default : PermOf n) = 1 := rfl
+
+instance : Unique (PermOf 0) := Unique.mk' _
+
+instance : Unique (PermOf 1) := Unique.mk' _
+
+end One
+
+instance : Inv (PermOf n) where
+  inv a := PermOf.mk
+    (toVector := a.invVector)
+    (invVector := a.toVector)
+    (getElem_toVector_lt := fun i hi => by
+      rcases a.getElem_surjective hi with ⟨j, hj, rfl⟩
+      exact hj.trans_eq' (a.getElem_invVector_getElem_toVector _))
+    (getElem_invVector_getElem_toVector := fun i hi => by
+      rcases a.getElem_surjective hi with ⟨j, hj, rfl⟩
+      congr 1
+      exact a.getElem_invVector_getElem_toVector _)
+
+section Inv
+
+@[simp]
+theorem getElem_inv_mk (a b : Vector ℕ n) {ha hab} (hi : i < n) :
+  (PermOf.mk a b ha hab)⁻¹[i] = b[i] := rfl
+
+@[simp]
+theorem getElem_invVector (hi : i < n) :
+  a.invVector[i] = a⁻¹[i] := rfl
+
+@[simp]
+theorem getElem_inv_getElem (hi : i < n) :
+    a⁻¹[a[i]] = i := a.getElem_invVector_getElem_toVector _
+
+@[simp]
+theorem getElem_getElem_inv (hi : i < n) :
+  a[a⁻¹[i]] = i := (a⁻¹).getElem_invVector_getElem_toVector _
+
+theorem eq_getElem_inv_iff (hi : i < n) (hj : j < n) :
+    i = a⁻¹[j] ↔ a[i] = j := by
+  rw [← (a⁻¹).getElem_inj (a.getElem_lt) hj, getElem_inv_getElem]
+
+theorem ne_getElem_inv_iff (hi : i < n) (hj : j < n) :
+    i ≠ a⁻¹[j] ↔ a[i] ≠ j := (a.eq_getElem_inv_iff _ _).ne
+
+theorem getElem_inv_eq_iff (hi : i < n) (hj : j < n) :
+    a⁻¹[i] = j ↔ i = a[j] := by
+  rw [← a.getElem_inj (a⁻¹.getElem_lt) hj, getElem_getElem_inv]
+
+theorem getElem_inv_ne_iff (hi : i < n) (hj : j < n) :
+    a⁻¹[i] ≠ j ↔ i ≠ a[j] := (a.getElem_inv_eq_iff _ _).ne
+
+end Inv
+
+instance : Mul (PermOf n) where
+  mul a b := {
+    toVector := Vector.ofFn (fun i => a.toVector[b[(i : ℕ)]])
+    invVector := Vector.ofFn (fun i => b⁻¹.toVector[a⁻¹[(i : ℕ)]])
+    getElem_toVector_lt := fun {i} hi => by
+      simp_rw [Vector.getElem_ofFn, getElem_toVector, getElem_lt]
+    getElem_invVector_getElem_toVector := fun {i} hi => by
+      simp_rw [Vector.getElem_ofFn, getElem_toVector, getElem_inv_getElem]}
+
+section Mul
+
+@[simp]
+theorem getElem_mul {i : ℕ} (hi : i < n) :
+    (a * b)[i] = a[b[i]] := Vector.getElem_ofFn _
+
+end Mul
+
+instance : SMul (PermOf n) ℕ where
```

Comment:
This action seems pretty weird to me; are the semantics of `PermOf` "this is a permutation of n elements", or "this is a permutation of the natural numbers that leaves all but the first n elements unchanged"? I'd argue this belongs as def with type `Perm (Fin n) ->* Perm Nat`

Reply:
It's the latter - "this is a permutation of the natural numbers that only moves the first n".

Reply:
https://github.com/linesthatinterlace/controlbits/blob/master/CBConcrete%2FCBConcrete%2FPermOf%2FMonoidHom.lean

I believe this repo is open but if you look at the `castSucc` and `castPred` maps here you can see that they have very nice definitions in terms of Vector.pop and Vector.push, which would not be possible in the same way with your suggestion. 

Again, I started with the perspective of "these represent permutations on `Fin n`" and it just isn't the right semantics because of the ways I found I naturally wanted to use them. One of the things I found over the course of my work was that bundling in propositions like this too early can cause real headaches.